src/Cli/dotnet/Commands/Package/Add/PackageAddCommand.cs

```diff
@@ -133,4 +139,214 @@ private string[] TransformArgs(PackageIdentityWithRange packageId, string tempDg
+            editor.Add(new CSharpDirective.Package(default) { Name = _packageId.Id, Version = version });
+            command.Directives = editor.Directives;
+            return (Save, Update);
+
+            void Save()
+            {
+                editor.SourceFile.Save();
+            }
+
+            void Update(string value)
+            {
+                // Update the C# file with the given version.
+                editor.Add(new CSharpDirective.Package(default) { Name = _packageId.Id, Version = value });
+                editor.SourceFile.Save();
+            }
+        }
+
+        (Action Revert, Action<string> Update, Action Save)? SetCentralVersion(string version)
+        {
+            // Find out whether CPM is enabled.
+            if (!MSBuildUtilities.ConvertStringToBool(projectInstance.GetProperty("ManagePackageVersionsCentrally")?.EvaluatedValue))
+            {
+                return null;
+            }
+
+            // Load the Directory.Packages.props project.
+            var directoryPackagesPropsPath = projectInstance.GetProperty("DirectoryPackagesPropsPath")?.EvaluatedValue;
+            if (!File.Exists(directoryPackagesPropsPath))
+            {
+                Reporter.Verbose.WriteLine($"Directory.Packages.props file does not exist: {directoryPackagesPropsPath}");
+                return null;
+            }
+
+            var snapshot = File.ReadAllText(directoryPackagesPropsPath);
+            var directoryPackagesPropsProject = projectCollection.LoadProject(directoryPackagesPropsPath);
+
+            const string packageVersionItemType = "PackageVersion";
+            const string versionAttributeName = "Version";
+
+            // Update existing PackageVersion if it exists.
+            var packageVersion = directoryPackagesPropsProject.GetItems(packageVersionItemType)
+                .LastOrDefault(i => string.Equals(i.EvaluatedInclude, _packageId.Id, StringComparison.OrdinalIgnoreCase));
+            if (packageVersion != null)
+            {
+                var packageVersionItemElement = packageVersion.Project.GetItemProvenance(packageVersion).LastOrDefault()?.ItemElement;
+                var versionAttribute = packageVersionItemElement?.Metadata.FirstOrDefault(i => i.Name.Equals(versionAttributeName, StringComparison.OrdinalIgnoreCase));
+                if (versionAttribute != null)
+                {
+                    versionAttribute.Value = version;
+                    directoryPackagesPropsProject.Save();
+
+                    // If user didn't specify a version and a version is already specified in Directory.Packages.props,
+                    // don't update the Directory.Packages.props (that's how the project-based equivalent behaves as well).
+                    if (!hasVersion)
+                    {
+                        skipUpdate = true;
+                        return (Revert: NoOp, Update: Unreachable, Save: Revert);
```

Comment:
I am finding the flow here pretty wonky. It feels like some of the cases update the file on disk before we know whether that's what we should really do, and other cases wait to confirm that's what we should do before doing the update. Does it have to be that way?

I was also kinda hoping that the case where we need to update a Directory.Packages.props file could share more with the project-based version of the command. I guess that is not practical to do.